### PR TITLE
Vue CLI 2 & 3 Cypress Code Coverage Scripts

### PR DESCRIPTION
**Pull request by kevinold**
For UNIFY-78

Associated PRs: #52 #55 

@cowboy Note that I've removed the symlink between`vue-cli-3` and `vue-cli-2` (2b973b7) because of differences in the files generated by each.

### CODE REVIEW

File: scripts/vue-cli-2/3.patch [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
++    HelloWorld,
++    Counter
+   }
+ }
+ </script>
+diff --git a/vue-cli-2/src/components/Counter.vue b/vue-cli-2/src/components/Counter.vue
+new file mode 100644
+index 0000000..5eb6269
+--- /dev/null
++++ b/vue-cli-2/src/components/Counter.vue
```
Comment (cowboy) [inline scripts/vue-cli-2/3.patch:136]: Missing the corresponding `Counter.spec.ct.js` spec file

**Review by cowboy — CHANGES_REQUESTED** [review]

**Review by cowboy — CHANGES_REQUESTED** [review]
x

File: scripts/vue-cli-3/3.patch [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
++
++  methods: {
++    increment() {
++      this.count++
++    },
++  },
++}
++</script>
+\ No newline at end of file
+diff --git a/vue-cli-3/src/components/CounterComp.spec.ct.js b/vue-cli-3/src/components/CounterComp.spec.ct.js
```
[the remark this answers is not in the document]
Comment (kevinold) [reply scripts/vue-cli-3/3.patch:168]: @cowboy Regarding the name of `CounterComp.vue` in the Vue 3 CLI example — it is using the Vue 3 Component syntax vs the traditional.  In this example I am shipping the Vue 2 syntax in `Counter.vue`, so wonder what your thoughts are for naming the Vue 3 component?

**Comment by kevinold** [comment]
@cowboy Updates made per notes and sync for Vue 2 and Vue 3 CLI's.  The associated PRs #52 #55 have been updated with the latest from the scripts.

File: scripts/vue-cli-2/pr.md [hunk 3]
```diff
@@ -10,10 +10,19 @@
    1. ${COMMAND_2}
    2. Add [cypress.json](cypress.json)
    3. Add [cypress/plugins/index.js](cypress/plugins/index.js)
-   4. Add [src/App.spec.ct.js](src/App.spec.ct.js), [src/components/HelloWorld.spec.ct.js](src/components/HelloWorld.spec.ct.js)
+   4. Add [src/App.spec.ct.js](src/App.spec.ct.js), [src/components/HelloWorld.spec.ct.js](src/components/HelloWorld.spec.ct.js), [src/Counter.vue](src/Counter.vue), [src/components/Counter.spec.ct.js](src/components/Counter.spec.ct.js)
```
Comment (cowboy) [inline scripts/vue-cli-2/pr.md:13]: Can you split this up into 2 lines, eg.

```suggestion
   1. Update [src/App.vue](src/App.vue) to reference new [src/components/Counter.vue](src/components/Counter.vue) component
   1. Add [src/App.spec.ct.js](src/App.spec.ct.js), [src/components/HelloWorld.spec.ct.js](src/components/HelloWorld.spec.ct.js), [src/components/Counter.spec.ct.js](src/components/Counter.spec.ct.js) spec files
```

Also, FWIW using `1.` for every bullet point means you won't ever have to re-number them.

File: scripts/vue-cli-2/pr.md [hunk 4]
```diff
@@ -10,10 +10,19 @@
...
    5. `npx cypress open-ct` (Notice that the fonts don't inherit global app styles)
    6. Edit [cypress/support/index.js](cypress/support/index.js) to import global app styles, the Cypress test preview should update automatically
    7. ${COMMIT_2}
+4. Add Cypress Code Coverage
+   1. ${COMMAND_3}
+   2. Edit [cypress.json](cypress.json) to enable `coverage`
+   3. Edit [cypress/plugins/index.js](cypress/plugins/index.js) to configure the Cypress code coverage task with Istanbul
+   4. Edit [cypress/support/commands.js](cypress/support/commands.js) to import Cypress code coverage support
```
Comment (cowboy) [inline scripts/vue-cli-2/pr.md:21]:
```suggestion
   4. Edit [cypress/support/index.js](cypress/support/index.js) to import Cypress code coverage support
```

File: scripts/vue-cli-2/pr.md [hunk 5]
```diff
@@ -10,10 +10,19 @@
...
+   5. Update [babel.config.js](babel.config.js) to conditionally use `babel-plugin-istanbul` when Cypress Component tests are running
+   6. Add [.nycrc](.nycrc)
```
Comment (cowboy) [inline scripts/vue-cli-2/pr.md:23]: Can you add a short bit of text to explain what this file is? Eg.

```suggestion
   6. Add [.nycrc](.nycrc) to configure istanbul
```

File: scripts/vue-cli-3/pr.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+   3. ${COMMIT_0}
+2. Update app to move global styles into main.css file
+   1. Move global app styles from [src/App.vue](src/App.vue) into [src/main.css](src/main.css)
+   2. Update [src/main.js](src/main.js)
+   3. ${COMMIT_1}
+3. Add Cypress component testing support with sample tests
+   1. ${COMMAND_2}
+   2. Add [cypress.json](cypress.json)
+   3. Add [cypress/plugins/index.js](cypress/plugins/index.js)
+   4. Add [src/App.spec.ct.js](src/App.spec.ct.js), [src/components/HelloWorld.spec.ct.js](src/components/HelloWorld.spec.ct.js), [src/Counter.vue](src/Counter.vue), [src/components/Counter.spec.ct.js](src/components/Counter.spec.ct.js)
```
Comment (cowboy) [inline scripts/vue-cli-3/pr.md:13]: Split into 2 lines?

```suggestion
   1. Update [src/App.vue](src/App.vue) to reference new [src/components/Counter.vue](src/components/Counter.vue) component
   1. Add [src/App.spec.ct.js](src/App.spec.ct.js), [src/components/HelloWorld.spec.ct.js](src/components/HelloWorld.spec.ct.js), [src/components/Counter.spec.ct.js](src/components/Counter.spec.ct.js) spec files
```

File: scripts/vue-cli-3/pr.md [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+   5. Add [src/CounterCompositionApi.vue](src/components/CounterCompositionApi.vue), a Composition API component and test in [src/components/CounterCompositionApi.spec.ct.js](src/components/CounterCompositionApi.spec.ct.js)
```
Comment (cowboy) [inline scripts/vue-cli-3/pr.md:14]:
```suggestion
   5. Add [src/CounterCompositionApi.vue](src/components/CounterCompositionApi.vue), a Composition API component with tests in [src/components/CounterCompositionApi.spec.ct.js](src/components/CounterCompositionApi.spec.ct.js)
```

File: scripts/vue-cli-3/pr.md [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+   6. `npx cypress open-ct` (Notice that the fonts don't inherit global app styles)
+   7. Edit [cypress/support/index.js](cypress/support/index.js) to import global app styles, the Cypress test preview should update automatically
+   8. ${COMMIT_2}
+4. Add Cypress Code Coverage
+   1. ${COMMAND_3}
+   2. Edit [cypress.json](cypress.json) to enable `coverage`
+   3. Edit [cypress/plugins/index.js](cypress/plugins/index.js) to configure the Cypress code coverage task with Istanbul
+   4. Edit [cypress/support/commands.js](cypress/support/commands.js) to import Cypress code coverage support
```
Comment (cowboy) [inline scripts/vue-cli-3/pr.md:22]:
```suggestion
   4. Edit [cypress/support/index.js](cypress/support/index.js) to import Cypress code coverage support
```

File: scripts/vue-cli-3/pr.md [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+   5. Update [babel.config.js](babel.config.js) to conditionally use `babel-plugin-istanbul` when Cypress Component tests are running
+   6. Add [.nycrc](.nycrc)
```
Comment (cowboy) [inline scripts/vue-cli-3/pr.md:24]: Same comment as the other readme

**Review by cowboy — CHANGES_REQUESTED** [review]

File: scripts/vue-cli-3.sh [hunk 10]
```diff
@@ -45,6 +45,11 @@ commit_all "Update app to move global styles into main.css file"
 step
 cmd yarn add -D cypress @cypress/vue@3 @cypress/webpack-dev-server webpack-dev-server
 apply_patch
-commit_all "Add Cypress with example component and page tests"
+commit_all "Add Cypress with example components and page tests"
+
+step
+cmd yarn add -D @cypress/code-coverage babel-plugin-istanbul
+apply_patch
+commit_all "Configure Cypress Code Coverage plugin and add additional component tests"
```
[the remark this answers is not in the document]
Comment (kevinold) [reply scripts/vue-cli-3.sh:53]: Good catch, adjusting and updating.

File: scripts/vue-cli-3/pr.md (hunk 9 above)
Comment (kevinold) [inline scripts/vue-cli-3/pr.md:25]:
```suggestion
   6. Add [.nycrc](.nycrc)  to configure istanbul
```

File: scripts/vue-cli-3.sh (hunk 10 above)
Comment (kevinold) [inline scripts/vue-cli-3.sh:53]:
```suggestion
commit_all "Configure Cypress Code Coverage plugin"
```

File: scripts/vue-cli-2.sh [hunk 11]
```diff
@@ -47,4 +47,9 @@ cmd yarn add -D cypress @cypress/vue @cypress/webpack-dev-server webpack-dev-ser
 apply_patch
 commit_all "Add Cypress with example component and page tests"
 
+step
+cmd yarn add -D @cypress/code-coverage babel-plugin-istanbul
+apply_patch
+commit_all "Configure Cypress Code Coverage plugin and add additional component tests"
```
Comment (kevinold) [inline scripts/vue-cli-2.sh:53]:
```suggestion
commit_all "Configure Cypress Code Coverage plugin"
```

**Comment by kevinold** [comment]
@cowboy Updates made per notes and sync for Vue 2 and Vue 3 CLI's. The associated PRs #52 #55 have been updated with the latest from the scripts.

**Review by cowboy — APPROVED** [review]
LGTM. One question: what if the vue-cli-3 example just had a composition API Counter component and not the old type, since the vue-cli-2 example has the old type or component? I'm also ok with this as-is, I just want to make sure we're keeping it simple wherever possible.

**Pull request merged**